numpy/_core/src/multiarray/common.h

```diff
@@ -172,13 +172,13 @@ check_and_adjust_axis(int *axis, int ndim)
  */
 #if (!defined __STDC_VERSION__ || __STDC_VERSION__ < 201112 \
      || (defined __GNUC__ && __GNUC__ < 4 + (__GNUC_MINOR__ < 9) \
-  && !defined __clang__) \
+     && !defined __clang__) \
      || (defined __clang__ && __clang_major__ < 8))
-# define NPY_ALIGNOF(type) offsetof(struct {char c; type v;}, v)
+#define NPY_ALIGNOF(type) __alignof__(type)
```

Comment:
What is the motivation to change these macro's? You don't seem to be using them anywhere either?

Reply:
![image](https://github.com/user-attachments/assets/301a4497-0274-4364-8791-0dccb05f772a)

Hi @r-devulap 
Change to the NPY_ALIGNOF macro, replacing `offsetof `with `__alignof__`, is motivated by the need for compliance with modern C++ standards and to eliminate compiler errors. The original use of offsetof with an anonymous struct may not be standard in C++11 and later, which leads to errors as shown in pic. 

On the other hand, modern compilers provide __alignof__, which is built-in, efficient, and compatible with both GCC and Clang, ensuring adherence to C++11 standards and simplifying alignment calculation

Reply:
1. On GCC and Clang (Linux/macOS), the `__alignof__` operator is supported and returns the required alignment of a type.
2. On MSVC,  `__alignof__` is not recognized, but the standard `alignof ` is used.

This causes an issue with failing of all Window CI.

Reply:
> The original use of offsetof with an anonymous struct may not be standard in C++11 and later, which leads to errors as shown in pic.

Does this happen on the main branch as well? I suggest we make a separate PR for this change cos it seems unrelated to this patch.

Reply:
> Does this happen on the main branch as well?

No